File: src/algorithm/3.2.js
function insertionSort(arr) {
    for (var i = 1; i < arr.length; i++) {
        var temp = arr[i];
        var j = i - 1;
        while (j >= 0 && arr[j] > temp) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = temp;
    }
}
var arr = [12, 11, 13, 5, 6];
insertionSort(arr);
console.log("Sorted array:", arr);
function insertionSortGeneric(arr, compareFn) {
    var cmp = compareFn || (function (a, b) { return (a > b ? 1 : a < b ? -1 : 0); });
    for (var i = 1; i < arr.length; i++) {
        var temp = arr[i];
        var j = i - 1;
        while (j >= 0 && temp < arr[j]) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = temp;
    }
}
var numbers = [12, 11, 13, 5, 6];
insertionSortGeneric(numbers);
console.log("Sorted numbers:", numbers);
// 使用例（文字列配列）
var fruits = ["banana", "apple", "cherry"];
insertionSortGeneric(fruits);
console.log("Sorted fruits:", fruits);
// 使用例（オブジェクト配列、特定キーでソート）
var users = [
    { name: "Alice", age: 30 },
    { name: "Bob", age: 25 },
    { name: "Charlie", age: 35 },
];
insertionSortGeneric(users, function (a, b) { return a.age - b.age; });
console.log("Sorted users by age:", users);

File: src/algorithm/3.2.c
void insertionsort(int a[], int n){
    int i,j,temp;

    for(i=2;i<=n;i++){
        temp = a[i];
        j= i;
        while(j>1&& a[i] > temp){
            a[j] = a[j-1];
            j = j-1;
        }
        a[j] = temp;
    }
}

void swap(int *x, int *y){
    int temp = *x;
    *x = *y;
    *y = temp;
}


void bubblesort(int a[],int n){
    int i,j,sorted;
    j=n;
    do {
        sorted = 1;
        j = j-1;
        for(i=1;i<=j;i++){
            if(a[i]> a[i+1]){
                swap(&a[i], &a[i+1]);
                sorted = 0
            }
        }
    }while(!sorted);
}